### DISCUSSION TITLE
Update Access-List via API - Add additional Rule

### DISCUSSION THREAD

**Discussion by Ian-Hartwell** [NginxProxyManager/nginx-proxy-manager#3195] [Q&A]
Hey,

Im trying to use the API to add an additional IP address to a Access-List, however the Update API overwrites the whole Clients section instead of just adding the extra one. 

Not a problem i thought, ill just go a fetch the existing list first and then append the new one! But the GET nginx/access-lists/ (both all and specific by id) doesnt return the full configuration, just a high level summary.

`{'id': 3, 'created_on': '2023-09-13 13:22:43', 'modified_on': '2023-09-13 13:22:43', 'owner_user_id': 1, 'name': 'LiveStream', 'meta': {}, 'satisfy_any': 0, 'pass_auth': 0, 'proxy_host_count': 0}`

Ive tried to read the Documentation (JSON files) but cant seem to see anywhere which gives me the data i need.

Am i being dim, or does the API not return this information anywhere?

Cheers

**Comment by gadget78** [comment]
i am also having this issue, 
it took me long enough just to get to this point, as there is no real documentation for the API :( 

ive been using the source code of the API here, to muddle my way though ... 
https://github.com/NginxProxyManager/nginx-proxy-manager/tree/develop/backend/routes/api

but im now at this point too, and i can confirm, it does seem to replace the whole list, and not "ADD" to it,
and i to cannot retrieve the list either, as like you state, it just gives a summary of the access-list instead 
so i also cannot find a work around :(

**Reply by aerickt** [reply]
Hey, would you happen to know a curl command to update the access list? I can't find any information on this.

**Comment by gadget78** [comment]
ok i re-visited this today, and i can confirm that you CAN retrieve the list of IPs from a access list !! 
but you have to specifically state PUT (and not GET), i found this kinda by accident lol..
for eg... 

`curl -X PUT -H "Content-Type: application/json; charset=UTF-8" -H "Authorization: Bearer toKeN..." http://192.168.1.9:81/api/nginx/access-lists/2`

will return the complete IP list contained within that access-list :)
```
{
  "id": 2,
  "created_on": "2024-03-21 00:06:06",
  "modified_on": "2024-03-27 21:35:44",
  "owner_user_id": 1,
  "name": "AccessList",
  "meta": {},
  "satisfy_any": 0,
  "pass_auth": 0,
  "proxy_host_count": 0,
  "owner": {
    "id": 1,
    "created_on": "2023-05-24 13:42:06",
    "modified_on": "2023-05-24 15:11:00",
    "is_deleted": 0,
    "is_disabled": 0,
    "email": "email@address",
    "name": "Administrator",
    "nickname": "Admin",
    "avatar": "",
    "roles": [
      "admin"
    ]
  },
  "items": [],
  "clients": [
    {
      "id": 18,
      "created_on": "2024-03-27 21:35:45",
      "modified_on": "2024-03-27 21:35:45",
      "access_list_id": 2,
      "address": "192.168.1.5",
      "directive": "allow",
      "meta": {}
    },
    {
      "id": 19,
      "created_on": "2024-03-27 21:35:45",
      "modified_on": "2024-03-27 21:35:45",
      "access_list_id": 2,
      "address": "192.168.1.6",
      "directive": "allow",
      "meta": {}
    }
  ],
  "proxy_hosts": []
}
```

**Reply by tumstech** [reply]
I am getting error response for this, stating payload is missing in PUT. Empty Put with "", or "{}" also not working

{"error":{"code":400,"message":"Payload is undefined"}}

**Comment by gadget78** [comment]
also found that the reason we cannot seemingly use the API to just "update" the list by sending one item,
its due to how its used within the GUI
1st - the GUI gets the existing list, and populates the GUI table,
2nd - you then add/remove items how you want them,
3rd -  you then click "save" on the GUI, this sends THAT complete LIST.. 
so the Update feature, creates the list based on the list it gets, so Adds all new entries, and removes all entries not in the payload... 

makes sense, when you understand how its using it ... 
so it DOES update any section you define, and leaves any sections you dont define...
